subprojects/language-native/src/main/java/org/gradle/language/cpp/internal/DefaultCppBinary.java

```diff
@@ -101,8 +94,13 @@ public DefaultCppBinary(Names names, ObjectFactory objects, Provider<String> bas
         nativeRuntime.getAttributes().attribute(MachineArchitecture.ARCHITECTURE_ATTRIBUTE, identity.getTargetMachine().getArchitecture());
         nativeRuntime.extendsFrom(getImplementationDependencies());
 
-        includePathConfiguration = includePathConfig;
-        includePath = componentHeaderDirs.plus(new FileCollectionAdapter(new IncludePath(includePathConfig)));
+        ArtifactView includeDirs = includePathConfiguration.getIncoming().artifactView(viewConfiguration -> {
+           viewConfiguration.attributes(attributeContainer -> {
+               attributeContainer.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.C_PLUS_PLUS_API));
```

Comment:
I've moved it up to the configuration and removed the view.

I had a bit of a chat today with @ljacomet and we determined that there is some wonkiness with the current implementation of artifact transforms in that if I specify the artifact type on both sides (i.e. on the published variant as well as on the consuming configuration) then variant selection fails because it does not take artifact transforms into account.  So for now, I've left the artifact type off of the published variant so that variant selection will succeed and then the artifact transform will take care of transforming to the directory type when the incoming configuration is resolved.

Reply:
I think it would be much better to have the correct attributes included in the published metadata and to keep the work around (using an artifact view) instead. So, I would keep the attribute on the outgoing variant and also keep the artifact view.

Can you also raise an issue for us to fix the problem so we can remove the work around?

Reply:
Fair enough.  I've changed it back and raised https://github.com/gradle/gradle/issues/8386.